### PR TITLE
renderer: reworked decode of BCn texture types

### PR DESCRIPTION
**Pull request by korenkonder**
- Added direct upload of BCn textures without decoding
- Reworked BCn decoder. Even though we don't need it, it's nice to have it for some cases

### CODE REVIEW

File: vita3k/renderer/src/gl/texture.cpp [hunk 1]
```diff
@@ -103,7 +103,7 @@ void configure_bound_texture(const renderer::TextureCacheState &state, const Sce
     const GLenum format = translate_format(base_format);
     const GLenum type = translate_type(base_format);
     const auto texture_type = gxm_texture.texture_type();
-    const bool is_swizzled = (texture_type == SCE_GXM_TEXTURE_SWIZZLED) || (texture_type == SCE_GXM_TEXTURE_CUBE) || (texture_type == SCE_GXM_TEXTURE_SWIZZLED_ARBITRARY) || (texture_type == SCE_GXM_TEXTURE_CUBE_ARBITRARY);
+    bool is_swizzled = (texture_type == SCE_GXM_TEXTURE_SWIZZLED) || (texture_type == SCE_GXM_TEXTURE_CUBE) || (texture_type == SCE_GXM_TEXTURE_SWIZZLED_ARBITRARY) || (texture_type == SCE_GXM_TEXTURE_CUBE_ARBITRARY);
```
Comment (Macdu) [inline vita3k/renderer/src/gl/texture.cpp:106]: why isn't it const anymore?

File: vita3k/renderer/src/gl/texture.cpp [hunk 2]
```diff
@@ -467,25 +615,27 @@ void upload_bound_texture(const SceGxmTexture &gxm_texture, const MemState &mem)
...
-
-        glPixelStorei(GL_UNPACK_ROW_LENGTH, static_cast<GLint>(pixels_per_stride));
+        if (block_compressed) {
+            glPixelStorei(GL_UNPACK_ROW_LENGTH, 0);
 
-        if (need_decompress_and_unswizzle_on_cpu)
-            glTexSubImage2D(upload_type, mip_index, 0, 0, width, height, GL_RGBA, GL_UNSIGNED_BYTE, pixels);
+            const GLenum format = translate_format(base_format);
+            size_t compressed_size = renderer::texture::get_compressed_size(base_format, width, height);
+            glCompressedTexSubImage2D(upload_type, mip_index, 0, 0, width, height, format, static_cast<GLsizei>(compressed_size), pixels);
```
Comment (Macdu) [inline vita3k/renderer/src/gl/texture.cpp:623]: Can you use glCompressedTexImage2D instead, Vita3K is still trying to be compatible with openGL 4.4 (the fact that we were using this function is a mistake).

File: vita3k/renderer/src/texture_format.cpp [hunk 3]
```diff
@@ -211,260 +214,322 @@ bool is_compressed_format(SceGxmTextureBaseFormat base_format, std::uint32_t wid
...
+    case SCE_GXM_TEXTURE_BASE_FORMAT_UBC4:
+    case SCE_GXM_TEXTURE_BASE_FORMAT_SBC4:
+        return ((width + 3) / 4) * ((height + 3) / 4) * 8;
+    case SCE_GXM_TEXTURE_BASE_FORMAT_UBC2:
+    case SCE_GXM_TEXTURE_BASE_FORMAT_UBC3:
+    case SCE_GXM_TEXTURE_BASE_FORMAT_UBC5:
+    case SCE_GXM_TEXTURE_BASE_FORMAT_SBC5:
+        return ((width + 3) / 4) * ((height + 3) / 4) * 16;
+    default:
+        return 0;
```
Comment (Macdu) [inline vita3k/renderer/src/texture_format.cpp:229]: Can you add a LOG_ERROR here ?

File: vita3k/renderer/src/gl/texture.cpp [hunk 4]
```diff
@@ -151,6 +154,53 @@ void configure_bound_texture(const renderer::TextureCacheState &state, const Sce
...
+ * \param width  Texture width.
+ * \param height Texture height.
+ *
+ * \return Void.
+ */
+static void remove_compressed_arbitrary_blocks(SceGxmTextureBaseFormat fmt, void *dest, const void *data, const std::uint32_t width, const std::uint32_t height) {
+    std::size_t w = (width + 3) / 4;
+    std::size_t h = (height + 3) / 4;
+
+    std::size_t a_w = nearest_power_of_two(width);
```
Comment (Macdu) [inline vita3k/renderer/src/gl/texture.cpp:172]: Can you rename this function to `next_power_of_two` (nearest is a bit misleading)?

File: vita3k/renderer/src/gl/texture.cpp (hunk 2 above)
Comment (korenkonder) [reply vita3k/renderer/src/gl/texture.cpp:622]: https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glCompressedTexSubImage2D.xhtml
`glCompressedTexSubImage2D` is compatible even with OpenGL 2.0 while `glCompressedTextureSubImage2D` isn't

File: vita3k/renderer/src/gl/texture.cpp (hunk 1 above)
Comment (korenkonder) [reply vita3k/renderer/src/gl/texture.cpp:106]: My bad

File: vita3k/renderer/src/gl/texture.cpp (hunk 2 above)
Comment (Macdu) [reply vita3k/renderer/src/gl/texture.cpp:622]: Oops my bad, misread the doc

File: vita3k/renderer/src/gl/texture.cpp (hunk 4 above)
Comment (korenkonder) [reply vita3k/renderer/src/gl/texture.cpp:172]: Done

**Pull request closed without merging**